### PR TITLE
docs: rethink recipe template

### PR DESCRIPTION
**Pull request by toph-allen**
Looking across all the recipes in the repo, almost all of them diverged from the template pretty significantly. The template was a little too prescriptive for the variety of structures that the recipes need. I adapted the lightweight format used by both the [R Cookbook](https://rc2e.com/) and [Python Cookbook](https://github.com/lpvcpp/learn_python/blob/master/D.%20Beazley%2C%20B.K.%20Jones%20-%20Python%20Cookbook%2C%203rd%20Edition.%202013.pdf) to preserve the important things that the existing template does well and address our specific needs.

- Problem: Some recipes differ between Python and R, either in required input parameters or more significantly in workflow.
  - Solution: Be more flexible about whether input and workflow descriptions happen outside or inside of language-specific tabs.
- Problem: Lack of consistency with headings. Some difficulty with flow in the template (e.g. Output described near the start, the heading "Recipe" being used to label the code snippets despite also being used to refer to the document itself).
  - Solution: Simplify headings (h2), following the conventions in the R and Python cookbooks from O'Reilly. Headings are "Problem", "Solution", Discussion [optional], See Also [optional]. Allow more flexibility with subheadings (h3...h*n*), depending on recipe requirements. Subheadings should be descriptive enough to understand what they're about from the ToC.
- Problem: Differences between example input variables.
  - Solution: Make some decisions. (1) Input variables that the user has to specify will be in all caps, other variables will be lower case. (2) We'll use actual examples of GUIDs, IDs, etc., where they are required by recipes, so that users know what shape of value they're looking for.

Here's what the new template looks like rendered:

Screen Shot 2024-06-26 at 03 34 09 PM@2x

### CODE REVIEW

File: _recipe-template.qmd [hunk 1]
```diff
@@ -1,76 +1,124 @@
 ---
-title: Title
+title: Recipe Title
 ---
 
-## Description
+## Problem
 
-When would you use this recipe?
+[Describe what the recipe does, or set up the problem the user can solve with it, e.g. "You need to assign tags to content on your Connect server."]
```
Comment (tdstein) [inline _recipe-template.qmd:7]: Would something like this also be acceptable?

```suggestion
[Describe what the recipe does or set up the problem the user can solve with it, e.g., "You need to assign tags to content on your Connect server" or "You want to organize content on Connect."]
```

File: _recipe-template.qmd [hunk 2]
```diff
@@ -1,76 +1,124 @@
...
 
-1.
+1. e.g. "The unique identifier (GUID) of the content you wish to modify." Link to other recipes if they generate the inputs, e.g. generating a tag.
 
 This recipe then does the following:
 
-1.
-
-## Recipe
+1. e.g. "Modifies the tags for the provided content item.
```
Comment (tdstein) [inline _recipe-template.qmd:19]: Should we use more natural language to describe this based on the solution? For example:

## Solution

Assign a tag to your content. This requires obtaining the content item's unique identifier. See [Getting a content identifier](#some-other-section).

Comment (tdstein) [reply _recipe-template.qmd:19]: Although, for the Python SDK the full example could by pass the need to lookup the GUID manually.

## Problem

You want to organize your content.

## Solution

Assign a tag to your content.

```python
from posit import connect

client = connect.Client() 
content = client.content.find_one(prefix="Your Content Name")
client.post(... yada-yada...
```

## Discussion

A tag is a simple way to organize your content. Connect tags support hierarchical groups... yada-yada...

File: _recipe-template.qmd [hunk 3]
```diff
@@ -1,76 +1,124 @@
...
 
 #### User-defined inputs ####
 # 1. Specify xyz. Any clarifying text and what the default value is (if any).
-xyz =
+ABC = "154bd2af-e8fa-4aa4-aab8-dcef701f4af9"
 # 2. Specify abc. Any clarifying text and what the default value is (if any).
-abc =
-###########################
+XYZ = "23"
+#############################
```
Comment (tdstein) [inline _recipe-template.qmd:45]: This doesn't look very polished, in my opinion. Ideally, we could provide links to other relevant parts of the Cookbook which show how to obtain the identifier. And then just use an arbitrary example in code block...

```python
from posit import connect

client = connect.Client()
content = client.content.get("154bd2af-e8fa-4aa4-aab8-dcef701f4af9")
content.update(description="My new description")
```

**Review by tdstein — COMMENTED** [review]
Thanks for putting this together @toph-allen. As we discussed earlier, I really like the Problem/Solution/Discussion structure. It provides structure but isn't overly prescriptive.

Here are some loosely held opinions to consider:

File: _recipe-template.qmd (hunk 2 above)
Comment (toph-allen) [reply _recipe-template.qmd:19]: Yeah, I think… we should probably leave it up to the recipe how to describe the requirements.

Like, if all they need is a GUID, you probably don't need an enumerated list.

Some recipes might do better with an enumerated list.

File: _recipe-template.qmd (hunk 3 above)
Comment (jonkeane) [reply _recipe-template.qmd:45]: This isn't a super strongly held belief / opinion, but I actually kinda like setting off the input values. It makes it clearer to (relative) novice users where the thing is that they need to edit / swap out for their own value

Comment (toph-allen) [reply _recipe-template.qmd:45]: I'm conflicted. I think I lean more towards Jon's point of view, but it's less important for some recipes.

I think removing the trailing line of pound signs makes this section look nicer. I go back and forth on making them all caps.

I think for longer recipes / more than one variable, the following is fine.

```python
#### User-defined inputs ####
# 1. Specify xyz. Any clarifying text and what the default value is (if any).
ABC = "154bd2af-e8fa-4aa4-aab8-dcef701f4af9"
# 2. Specify abc. Any clarifying text and what the default value is (if any).
XYZ = "23"
```

For shorter recipes, especially with one variable, the following is probably fine.

```r
# Define the content guid
GUID <- ""154bd2af-e8fa-4aa4-aab8-dcef701f4af9"
```

File: _recipe-template.qmd (hunk 2 above)
Comment (toph-allen) [reply _recipe-template.qmd:19]: I think if we remove input lists in the prose, let's leave them in (at least longer) recipes.

File: _recipe-template.qmd (hunk 3 above)
Comment (toph-allen) [reply _recipe-template.qmd:45]: Like, if you're just specifying a tag name, you don't need

#### User-defined inputs ####
# 1. The tag name.
TAG_NAME = "Marketing"
#############################

Comment (tdstein) [reply _recipe-template.qmd:45]: Yeah, I think it's recipe dependent.

@toph-allen - for the casing, I've been treating them as Python constants, which are conventionally UPPER_SNAKE_CASE.

File: _recipe-template.qmd (hunk 2 above)
Comment (tdstein) [reply _recipe-template.qmd:19]: > Like, if all they need is a GUID, you probably don't need an enumerated list.

This was the part I was getting hung up on. That makes sense to me.

File: _recipe-template.qmd [hunk 4]
```diff
@@ -1,76 +1,124 @@
...
```
Comment (tdstein) [inline _recipe-template.qmd:45]: Let's get rid of the `#####` stuff. The constants should be in a predictable location. In Python that will be directly after imports and before implementation code.

```python
from posit import connect

CONTENT_GUID = ...
BUNDLE_ID = ...

client = connect.Client()
content = client.content.get(CONTENT_GUID)
bundle = content.bundles.get(BUNDLE_ID)
```

Comment (toph-allen) [reply _recipe-template.qmd:45]: With optional comments, maybe?

```python
from posit import connect

client = connect.Client()

# Get the abc from the dashboard, or see above for more info.
ABC = "154bd2af-e8fa-4aa4-aab8-dcef701f4af9"
# The ID of the xyz you wish to do something to.
XYZ = "23"

new_var = do_something(ABC, XYZ)
output_df = function(ABC, new_var)
```

For now we can remove the `###` and numbers, and if it makes sense to add them back it won't be hard work.

File: _recipe-template.qmd (hunk 3 above)
Comment (toph-allen) [reply _recipe-template.qmd:45]: Yeah, in R there aren't conventions around… well, most things, but I think using capitals in our recipes for user-defined consts is great.

**Review by tdstein — APPROVED** [review]
❤️

**Pull request merged**